notifier/notify_windows.go

```diff
@@ -2,351 +2,342 @@ package main
-	case "buttonClicked":
-		for _, actions := range n.AvailableActions {
-			if actions.Text == data.button {
-				n.SelectAction(actions.ID)
-				return
-			}
+	// Lock library
+	lib.Lock()
+	defer lib.Unlock()
+
+	title, _ := windows.UTF16PtrFromString(n.Title)
+	message, _ := windows.UTF16PtrFromString(n.Message)
+	titleP := unsafe.Pointer(title)
+	messageP := unsafe.Pointer(message)
+	if lib != nil {
+		// Create new notification object
+		notificationTemplate, _, err := lib.createNotification.Call(uintptr(titleP), uintptr(messageP))
+		if notificationTemplate == 0 {
+			log.Errorf("notify: failed to create notification: %q", err)
 		}
+		// Make sure memory is freed when done
+		defer func() { _, _, _ = lib.deleteNotification.Call(notificationTemplate) }()
 
-		log.Warningf("failed to handle button click: button %s is reported to have been clicked but is no registered button. Available options: %+v", data.button, n.AvailableActions)
-	default:
-		log.Debugf("failed to handle SnoreToast-Action %s: not implemented yet", data.action)
-	}
-}
-
-// Snoretoast helper-Functions:
-
-// Verifies that the Argument contians no Semicolon which would be difficult (in some cases impossible) to parse in a Pipe-Response
-func verifySnoreToastArgumentSyntax(arg string) error {
-	if strings.Contains(arg, ";") {
-		return fmt.Errorf("the SnoreToast-Argument %s would contain a semicolon which would screw up the pipe responses", arg)
-	}
-	return nil
-}
-
-func parseSnoreToastActionCallback(in string) actionCallback {
-	ret := actionCallback{}
-
-	for _, elem := range strings.Split(in, ";") {
-		if elem == "" {
-			continue
+		// Set Portmaster icon.
+		iconLocation, err := ensureAppIcon()
+		if err != nil {
+			log.Warningf("notify: failed to write icon: %s", err)
 		}
-
-		elemSplit := strings.SplitN(elem, "=", 2)
-		if len(elemSplit) != 2 {
-			log.Warningf("failing to parse snoretoast-Response %s into Key=Value-Pair; Response: %s", elem, in)
-			continue
+		iconPathUTF, _ := windows.UTF16PtrFromString(iconLocation)
+		iconPathP := unsafe.Pointer(iconPathUTF)
+		_, _, _ = lib.setImage.Call(notificationTemplate, uintptr(iconPathP))
+
+		// Set all the required actions
+		for _, action := range n.AvailableActions {
+			textUTF, _ := windows.UTF16PtrFromString(action.Text)
+			textP := unsafe.Pointer(textUTF)
+			_, _, _ = lib.addButton.Call(notificationTemplate, uintptr(textP))
 		}
 
-		switch elemSplit[0] {
-		case "action":
-			ret.action = elemSplit[1]
-		case "button":
-			ret.button = elemSplit[1]
-		case "notificationId":
-			ret.id = elemSplit[1]
-		case "version", "pipe": // not needed: do nothing
-		default:
-			log.Infof("failed to parse key %s from SnoreToast-Response into struct: receivedkey is unknown to Portmaster (%s)", elemSplit[0], elem)
+		// Show notification and delete c notification object
+		rc, _, err := lib.showNotification.Call(notificationTemplate)
+		if int64(rc) == -1 {
+			log.Errorf("notify: failed to show notification: %q", err)
```

Comment:
If you expect a weird error message, using `%q` is good, but it is generic, just using `%s` makes it easier to read.